pepper-apis/src/main/java/org/broadinstitute/ddp/route/PatchLastVisitedActivitySectionRoute.java

```diff
@@ -0,0 +1,77 @@
+import org.broadinstitute.ddp.db.TransactionWrapper;
+import org.broadinstitute.ddp.db.dao.ActivityInstanceDao;
+import org.broadinstitute.ddp.db.dto.ActivityInstanceDto;
+import org.broadinstitute.ddp.db.dto.LanguageDto;
+import org.broadinstitute.ddp.model.activity.instance.ActivityInstance;
+import org.broadinstitute.ddp.service.ActivityInstanceService;
+import org.broadinstitute.ddp.util.ResponseUtil;
+import org.broadinstitute.ddp.util.RouteUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import spark.Request;
+import spark.Response;
+import spark.Route;
+
+
+public class PatchLastVisitedActivitySectionRoute implements Route {
+
+    private static final Logger LOG = LoggerFactory.getLogger(PatchLastVisitedActivitySectionRoute.class);
+    private ActivityInstanceService actInstService;
+
+    public PatchLastVisitedActivitySectionRoute(ActivityInstanceService actInstService) {
+        this.actInstService = actInstService;
+    }
+
+    @Override
+    public Object handle(Request request, Response response) {
+        String userGuid = request.params(RouteConstants.PathParam.USER_GUID);
+        String studyGuid = request.params(RouteConstants.PathParam.STUDY_GUID);
+        String instanceGuid = request.params(RouteConstants.PathParam.INSTANCE_GUID);
+        JsonElement data = new JsonParser().parse(request.body());
+
+        LOG.info("Request to update last visited section on instance {} for participant {} in study {}",
+                instanceGuid, userGuid, studyGuid);
+
+        if (!data.isJsonObject() || data.getAsJsonObject().entrySet().size() == 0) {
+            ResponseUtil.halt400ErrorResponse(response, ErrorCodes.MISSING_BODY);
+        }
+
+        TransactionWrapper.useTxn(handle -> {
+            JsonObject payload = data.getAsJsonObject();
+            JsonElement lastVisitedActivitySectionValue = payload.get("lastVisitedActivitySection");
+            int lastVisitedActivitySection = lastVisitedActivitySectionValue.getAsInt();
+            ActivityInstanceDto instanceDto = RouteUtil.findAccessibleInstanceOrHalt(
+                    response, handle, userGuid, studyGuid, instanceGuid);
+
+            LanguageDto preferredUserLanguage = RouteUtil.getUserLanguage(request);
+            String isoLangCode = preferredUserLanguage.getIsoCode();
+
+            ContentStyle style = RouteUtil.parseContentStyleHeaderOrHalt(request, response, ContentStyle.STANDARD);
+            LOG.info("Using ddp content style {} to format activity content", style);
+
+            LOG.info("Attempting to find a translation for the following language: {}", isoLangCode);
+            Optional<ActivityInstance> inst = actInstService.getTranslatedActivity(
+                    handle, userGuid, instanceDto.getActivityType(), instanceGuid, isoLangCode, style
+            );
+
+            if (!inst.isPresent()) {
+                response.status(HttpStatus.SC_NOT_FOUND);
+            }
```

Comment:
I don't see the point of this whole block of code. I think it's unnecessary for this route to do its work. The `RouteUtil.findAccessibleInstanceOrHalt()` call will have already determined if activity instance exists or not and return `404`.

Reply:
if I understand correctly, I can just write
```suggestion
            TransactionWrapper.useTxn(handle -> {
            int lastVisitedSection = dataObject.getLastVisitedSection();
            ActivityInstanceDto instanceDto = RouteUtil.findAccessibleInstanceOrHalt(
                    response, handle, userGuid, studyGuid, instanceGuid);

            handle.attach(ActivityInstanceDao.class)
                    .updateLastVisitedSectionByInstanceGuid(instanceGuid, lastVisitedSection);
        });
```
and that's it.
Am I right?

Reply:
I think you need the `RouteUtil.findAccessibleInstanceOrHalt()` in there as well. And we'll need `404` since that's one of the things `RouteUtil.findAccessibleInstanceOrHalt()` returns.